### ISSUE TITLE
Can't narrow a type of a variable of TypeVar type with `type(...) is ...`

### ISSUE THREAD

**Issue by jack1142** [microsoft/pyright#3879]
**Describe the bug**
It doesn't seem to be possible to further narrow a type of a variable if it's a TypeVar.

**To Reproduce**
Steps to reproduce the behavior.
1. Put code from "Screenshots or Code" into a file.
2. Run `pyright` on it.
3. See results:
```
❯ pyright repro.py 
No configuration file found.
pyproject.toml file found at /home/ubuntu/work/label-doconly-changes.
Loading pyproject.toml file at /home/ubuntu/work/label-doconly-changes/pyproject.toml
Pyproject file "/home/ubuntu/work/label-doconly-changes/pyproject.toml" is missing "[tool.pyright]" section.
stubPath /home/ubuntu/work/label-doconly-changes/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
pyright 1.1.268
/home/ubuntu/work/label-doconly-changes/repro.py
  /home/ubuntu/work/label-doconly-changes/repro.py:23:21 - error: Argument of type "NodeT@func" cannot be assigned to parameter "foo" of type "Foo" in function "accepts_foo"
    "Node*" is incompatible with "Foo" (reportGeneralTypeIssues)
  /home/ubuntu/work/label-doconly-changes/repro.py:24:21 - error: Argument of type "NodeT@func" cannot be assigned to parameter "foo" of type "Foo" in function "accepts_foo"
    "Node*" is incompatible with "Foo" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations 
Completed in 0.656sec
```

**Expected behavior**
I expected no errors.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
from typing import TypeVar

class Node:
    ...

NodeT = TypeVar("NodeT", bound=Node)

class Foo(Node):
    ...

class Bar(Node):
    ...

def accepts_foo(foo: Foo) -> None:
    ...

def returns_node() -> Node:
    return Foo()


def func(a: NodeT, b: NodeT) -> None:
    if type(a) is Foo:
        accepts_foo(a)
        accepts_foo(b)
```

**VS Code extension or command-line**
I'm running pyright 1.1.268 from command-line

**Additional context**
Replacing `func()` with:
```py
def func(a: NodeT, b: NodeT) -> None:
    if isinstance(a, Foo):
        accepts_foo(a)
        accepts_foo(b)
```
narrows down `a` successfully but `b` doesn't get narrowed:
```
❯ pyright repro.py
No configuration file found.
pyproject.toml file found at /home/ubuntu/work/label-doconly-changes.
Loading pyproject.toml file at /home/ubuntu/work/label-doconly-changes/pyproject.toml
Pyproject file "/home/ubuntu/work/label-doconly-changes/pyproject.toml" is missing "[tool.pyright]" section.
stubPath /home/ubuntu/work/label-doconly-changes/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
pyright 1.1.268
/home/ubuntu/work/label-doconly-changes/repro.py
  /home/ubuntu/work/label-doconly-changes/repro.py:24:21 - error: Argument of type "NodeT@func" cannot be assigned to parameter "foo" of type "Foo" in function "accepts_foo"
    "Node*" is incompatible with "Foo" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
Completed in 0.646sec
```

**Comment by erictraut** [comment]
Yes, this is expected. Type narrowing applies only to the variable or expression within the check. It doesn't narrow other variables or expressions.

**Comment by erictraut** [comment]
Is there a reason you don't want to use `isinstance` here? The logic for that type narrowing pattern supports this case.

**Comment by erictraut** [comment]
I've enhanced the `type(x) is y` type guard logic to handle the case where `x` is defined as a bound TypeVar. This will be included in the next release.

Here's the [commit](https://github.com/microsoft/pyright/commit/d6d188b3972d0f81294d7cb9e4a2bb4f195226a1).

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.269, which I just published. It will also be included in a future version of pylance.
